### ISSUE TITLE
AttributeError: 'numpy.ndarray' object has no attribute 'extend'

### ISSUE THREAD

**Issue by nancynanxiaodan** [santoshphilip/eppy#96]
Hi all,

I got an error saying that 'numpy.ndarray' object(which is the schedule in the following code) has no attribute 'extend'.

    for i in range(7):
        for schedulename in range(len(ScheduleCompact)):
            if ScheduleCompact[schedulename].Name == OccupancySchedule[i][0]:
                Schedule = ScheduleCompact.pop(schedulename)

                p = len(OccupancySchedule[i]) - 1 + 5
                if len(Schedule['obj']) < p:
                    number_extra_array_space = p-len(Schedule['obj'])
                    extra_array_space=['' for m in range(number_extra_array_space)]
                    print Schedule
                    Schedule['obj'].extend(extra_array_space)

What I am trying to do here is I have a new occupancy schedule of one compact schedule in energyplus, I would like to use Eppy to locate this origin compact schedule and then replace it with my new occupancy schedule. However the length of these 2 schedule will probably be different, so if the new occupancy schedule is longer than the original one, therefore, I need to extend the origin schedule to be the same size, and then replace the content. 

However, I could not using extend function. The strange thing here is that I am able to change it in a standalone mode, which I provide the code below, this is working probably.

    from eppy.modeleditor import IDF

    iddfile = "C:\\Users\\nancy\\PycharmProjects\\occupancy_schedule\\iddfiles\\Energy+V8_4_0.idd"
    fname1 = "C:\\Users\\nancy\\PycharmProjects\\occupancy_schedule\\idffiles\\CH.idf"

    IDF.setiddname(iddfile)
    house = IDF(fname1)

    ScheduleCompact = house.idfobjects['SCHEDULE:COMPACT']

    schedule = ScheduleCompact.pop(2)

    a = ['', '']
    schedule['obj'].extend(a)
    print schedule

I am not quiet sure, what did I do wrong here.
May ask for help please?

Many thanks
Nancy

**Comment by jamiebull1** [comment]
I don't think the schedule in your second example is an `numpy.ndarray`. What you're `extend`ing there is just a list.

It's a bit confusing to read your code since you use CapitalCase for variables. Usually in Python this is only used for classes. I've suggested a couple of things that might be the problem though.


```
for i in range(7):
    for schedulename in range(len(ScheduleCompact)):  # schedulename is just an integer here?
        if ScheduleCompact[schedulename].Name == OccupancySchedule[i][0]:  # ScheduleCompact is a list?
            Schedule = ScheduleCompact.pop(schedulename)  # you're popping from a list that you're still iterating over which can lead to unexpected behaviour

            p = len(OccupancySchedule[i]) - 1 + 5
            if len(Schedule['obj']) < p:
                number_extra_array_space = p-len(Schedule['obj'])
                extra_array_space=['' for m in range(number_extra_array_space)]
                print Schedule
                Schedule['obj'].extend(extra_array_space)
```

If the code is doing what I think it is, this might be closer to what you want to do:

```
from eppy.modeleditor import extendlist

compact_schedules = house.idfobjects['SCHEDULE:COMPACT']

for i in range(7):
    for schedule in compact_schedules:
        if schedule.Name == OccupancySchedule[i][0]:  # not sure what OccupancySchedule is here
            p = len(OccupancySchedule[i]) - 1 + 5
            extendlist(schedule['obj'], p, value='')
            print schedule
```

Let me know if this helps.

**Comment by nancynanxiaodan** [comment]
Hey Jamie,

Many thx for your reply. I will take care of the name of variables, many thanks for noticing me about this.

    for i in range(7):
            for schedulename in range(len(ScheduleCompact)): # schedulename is an integer
                if ScheduleCompact[schedulename].Name == OccupancySchedule[i][0]: # schedulecompact is a list contains all the idf objects 'SCHEDULE:COMPACT', and the OccupancySchedule is a list contains the new occupancy schedules which will replace the old ones in schedulecompact
                    Schedule = ScheduleCompact.pop(schedulename)

example of occupancyschedule:
` ['Bedroom1_occ', 'Until: 06:10', 1, 'Until: 23:30', 0, 'Until: 24:00', 1]`

I think the built-in function could fulfill what I required. I will check it within my code now, and will let you know shortly.

Do appericate your kind help.

Thanks,
Nancy

**Comment by nancynanxiaodan** [comment]
Hey Jamie,

I've tried the buit-in function of extend, and the error I am getting is the same, however, I tried to use it standalone, it works. 

    for i in range(len(ScheduleCompact)):
        if ScheduleCompact[i].Name == 'Bathroom_occ':
            schedule = ScheduleCompact[i]
            a = 45
            extendlist(schedule['obj'], a, value='')
            break

I've pasted my part code below:


    for schedulename in range(len(ScheduleCompact)):
            if ScheduleCompact[schedulename].Name == OccupancySchedule[i][0]:
                schedule = ScheduleCompact.pop(schedulename)

                p = len(OccupancySchedule[i]) - 1 + 5
                if len(schedule['obj']) < p:
                    extendlist(schedule['obj'], p, value='')
                    print schedule['obj']

And, I am sharing my whole code in the following link: [https://www.dropbox.com/s/kfpvo3oxhed30t6/test_occ_schedule.py?dl=0](url)
And the idf file:
[https://www.dropbox.com/s/42fv76296wzbhqr/CH.idf?dl=0](url)

Basically, what I am trying to do here is, that every time a new occupancy schedule generated will be assigned to one IDF file and save the IDF file with a new name, and then repeat the same process, in order to have occupants based IDF file for EP+ to run.

Any tips, many thanks!
Nancy

**Comment by jamiebull1** [comment]
I can't run your code to test what's going on without the .dat files.

Do you know how to use the debugger in Pycharm? I suggest you set a breakpoint and step through the code in both the standalone version that works, and in the version that doesn't, and make sure that the variable names are referring to what you expect them to. That should give you something to go on.

**Comment by nancynanxiaodan** [comment]
Yes, I did debug these 2 code separately, and the variables names are correct and have the same attributes. 

Sorry for the missing files, they are here:
https://www.dropbox.com/sh/ys9n0n9xtv8gt17/AADAKELwb2y6dHElBep22Wloa?dl=0

I am still debugging right now, and still cant figure it our..

Any tips are really appreciated.
Nancy

**Comment by jamiebull1** [comment]
The problem is that you're changing the `list` into an `np.ndarray` if it is too long.

This then breaks when you try to extend the list later.

This is how I would reformat your code, and it seems to run fine now. You can check if it does what you expect.

```python
from itertools import product
# change idf based on new occupancy schedule, and save to individual idf file
for occ_sch, schedule in product(OccupancySchedule, ScheduleCompact):  # more pythonic than nested for loops
    if schedule.Name == occ_sch[0]:
        p = len(occ_sch) - 1 + 5
        # if schedule is shorter than required then extend it
        if len(schedule['obj']) < p:
            extendlist(schedule['obj'], p, value='')
        # if schedule is longer than required keep popping one off the end
        while len(schedule['obj']) > p:
            schedule['obj'].pop()
    
        j = 5
        for w in range(1,len(occ_sch)):
            schedule['obj'][j] = occ_sch[w]
            j=j+1
```

**Comment by nancynanxiaodan** [comment]
Hey Jamie, 

Many thanks! I am going to check this out later today!

Thanks very much for your kind help! Do appericate!
Nancy

**Comment by jamiebull1** [comment]
Hi @nancynanxiaodan ,

Is this working for you? Can we close the issue?

Jamie

**Comment by nancynanxiaodan** [comment]
Hey Jamie,
I've fix this problem! many thx! yes it can be closed now :)

**Comment by jamiebull1** [comment]
That's great. Glad it worked for you!
